Librairies python : pandas et numpy   
Numpy pour générer un tableau de deux entiers aléatoires (0 ou 1) de manière équiprobable   
Pandas pour rendre les données plus lisibles avec des dataframe et noms de colonnes.

In [73]:
import numpy as np
import pandas as pd

On génère 10000 couples avec leurs deux enfants.   
On peut donc avoir soit (Boy, Boy), soit (Girl, Girl) soit (Boy, Girl) soit (Girl, Boy)

In [74]:
children = pd.DataFrame(np.random.randint(0, 2, size=(10000, 2)), columns=['FirstChild', 'SecondChild'])

Les données sont rendues plus lisibles en transformant les 0 en garcon et les 1 en fille

In [75]:
children['FirstChild'] = children['FirstChild'].apply(lambda x: 'Boy' if x == 0 else 'Girl')
children['SecondChild'] = children['SecondChild'].apply(lambda x: 'Boy' if x == 0 else 'Girl')
children

FirstChild SecondChild
0           Boy         Boy
1           Boy         Boy
2           Boy         Boy
3           Boy         Boy
4           Boy         Boy
5          Girl         Boy
6           Boy         Boy
7           Boy         Boy
8          Girl        Girl
9          Girl        Girl
10         Girl        Girl
11         Girl        Girl
12         Girl         Boy
13          Boy         Boy
14          Boy        Girl
15         Girl        Girl
16         Girl        Girl
17         Girl         Boy
18          Boy        Girl
19          Boy         Boy
20          Boy        Girl
21         Girl         Boy
22          Boy        Girl
23          Boy         Boy
24         Girl         Boy
25          Boy         Boy
26         Girl        Girl
27          Boy         Boy
28         Girl         Boy
29          Boy        Girl
...         ...         ...
9970        Boy         Boy
9971        Boy        Girl
9972        Boy         Boy
9973        Boy         Boy
9974       Girl         Boy
9975       Girl         Boy
9976       Girl        Girl
9977        Boy         Boy
9978       Girl        Girl
9979        Boy        Girl
9980       Girl         Boy
9981       Girl        Girl
9982        Boy        Girl
9983       Girl        Girl
9984        Boy        Girl
9985        Boy        Girl
9986       Girl         Boy
9987       Girl         Boy
9988        Boy        Girl
9989        Boy        Girl
9990        Boy         Boy
9991        Boy        Girl
9992       Girl        Girl
9993        Boy        Girl
9994        Boy        Girl
9995        Boy         Boy
9996       Girl        Girl
9997        Boy        Girl
9998       Girl         Boy
9999       Girl        Girl

[10000 rows x 2 columns]

Seuls les couples ayant au moins un garcon nous intéresse.   
La condition logique est donc : le premier enfant est un garcon OU le second est un garcon.    
Ce n'est pas un OU exclusif, les deux enfants peuvent être des garcons.

In [76]:
at_least_one_boy = children[(children['FirstChild'] == 'Boy') | (children['SecondChild'] == 'Boy')]
at_least_one_boy.reset_index()

index FirstChild SecondChild
0         0        Boy         Boy
1         1        Boy         Boy
2         2        Boy         Boy
3         3        Boy         Boy
4         4        Boy         Boy
5         5       Girl         Boy
6         6        Boy         Boy
7         7        Boy         Boy
8        12       Girl         Boy
9        13        Boy         Boy
10       14        Boy        Girl
11       17       Girl         Boy
12       18        Boy        Girl
13       19        Boy         Boy
14       20        Boy        Girl
15       21       Girl         Boy
16       22        Boy        Girl
17       23        Boy         Boy
18       24       Girl         Boy
19       25        Boy         Boy
20       27        Boy         Boy
21       28       Girl         Boy
22       29        Boy        Girl
23       31       Girl         Boy
24       32       Girl         Boy
25       33       Girl         Boy
26       35        Boy         Boy
27       36        Boy         Boy
28       38       Girl         Boy
29       39        Boy        Girl
...     ...        ...         ...
7454   9962        Boy        Girl
7455   9963       Girl         Boy
7456   9965        Boy         Boy
7457   9966       Girl         Boy
7458   9967        Boy        Girl
7459   9968        Boy         Boy
7460   9969        Boy         Boy
7461   9970        Boy         Boy
7462   9971        Boy        Girl
7463   9972        Boy         Boy
7464   9973        Boy         Boy
7465   9974       Girl         Boy
7466   9975       Girl         Boy
7467   9977        Boy         Boy
7468   9979        Boy        Girl
7469   9980       Girl         Boy
7470   9982        Boy        Girl
7471   9984        Boy        Girl
7472   9985        Boy        Girl
7473   9986       Girl         Boy
7474   9987       Girl         Boy
7475   9988        Boy        Girl
7476   9989        Boy        Girl
7477   9990        Boy         Boy
7478   9991        Boy        Girl
7479   9993        Boy        Girl
7480   9994        Boy        Girl
7481   9995        Boy         Boy
7482   9997        Boy        Girl
7483   9998       Girl         Boy

[7484 rows x 3 columns]

Sur 10000 couples, 7484 ont au moins un garcon. On s'intéresse maintenant à la proportion de ceux qui ont deux enfant parmi ces 7484 couples.

In [78]:
two_boys_freq = at_least_one_boy[at_least_one_boy['FirstChild'] == at_least_one_boy['SecondChild']].shape[0]/at_least_one_boy.shape[0]
print("Proportion des couples ayant 2 garcons sachant que l'un des deux enfants est un garcon", two_boys_freq)

Proportion des couples ayant 2 garcons sachant que l'un des deux enfants est un garcon 0.32950293960448956


Parmi les couples ayant deux enfants dont l'un d'eux est un garcon, 32.95% est la proportion de ceux qui ont deux enfants.   
Le sondage de Michael Launay qui demandait aux individus en regardant autour d'eux de une personne ayant deux enfant dont l'un d'eux est un garcon si l'autre enfant est aussi un garcon donne le résultat suivant:   
42% != 32.95%   


L'une des raisons qui pourrait expliquer cette nette différence, vient peut être du fait que dans le cas réel, beaucoup de couple ont tendance de faire plus de garcon que de fille et vice versa.   
Si bien  